### ISSUE TITLE
Slow histogram request

### ISSUE THREAD

**Issue by hc4** [Graylog2/graylog2-server#4051]
I've noticed slow elastic queries.
After digging it down I found that problem is caused by not otimal histogramm request in case of empty query.

Current graylog histogramm query:
```
{
  "query" : {
    "query_string" : {
      "query" : "*",
      "allow_leading_wildcard" : false
    }
  },
  "aggregations" : {
    "gl2_filter" : {
      "filter" : {
        "bool" : {
          "must" : [ {
            "range" : {
              "timestamp" : {
                "from" : "2017-08-02 17:59:46.786",
                "to" : "2017-08-02 18:04:46.786",
                "include_lower" : true,
                "include_upper" : true
              }
            }
          }, {
            "query_string" : {
              "query" : "streams:573c6b2e1d8ed80b7a82fcb2"
            }
          } ]
        }
      },
      "aggregations" : {
        "gl2_histogram" : {
          "date_histogram" : {
            "field" : "timestamp",
            "interval" : "1m"
          }
        }
      }
    }
  }
}
```

Optimized query:
```
{
  "query" : {
    "bool" : {
      "must" : {
        "match_all" : { }
      },
	  "filter" : {
        "bool" : {
          "must" : [ {
            "range" : {
              "timestamp" : {
                "from" : "2017-08-02 17:59:46.786",
                "to" : "2017-08-02 18:04:46.786",
                "include_lower" : true,
                "include_upper" : true
              }
            }
          }, {
            "query_string" : {
              "query" : "streams:573c6b2e1d8ed80b7a82fcb2"
            }
          } ]
        }
      }
    }
  },
  "aggregations" : {
    "gl2_filter" : {
      "filter" : {
        "match_all" : { }
      },
      "aggregations" : {
        "gl2_histogram" : {
          "date_histogram" : {
            "field" : "timestamp",
            "interval" : "1m"
          }
        }
      }
    }
  }
}
```

First version tooks about 1m in my setup, second version less than 1s :)
Intresting thing is that data request uses such optimization.

## Your Environment

* Graylog Version: 2.3.0
* Elasticsearch Version: 5.5.1

**Comment by dennisoelkers** [comment]
Thanks for your valuable input @hc4 (as usual)! We will try that out and see if it makes sense for us to adapt our query.

**Comment by hc4** [comment]
Actually empty filtter in aggs sections is not needed.
So request should look like this:
```
{
  "query" : {
    "bool" : {
      "must" : {
        "match_all" : { }
      },
	  "filter" : {
        "bool" : {
          "must" : [ {
            "range" : {
              "timestamp" : {
                "from" : "2017-08-02 17:59:46.786",
                "to" : "2017-08-02 18:04:46.786",
                "include_lower" : true,
                "include_upper" : true
              }
            }
          }, {
            "query_string" : {
              "query" : "streams:573c6b2e1d8ed80b7a82fcb2"
            }
          } ]
        }
      }
    }
  },
  "aggregations" : {
    "gl2_histogram" : {
      "date_histogram" : {
        "field" : "timestamp",
        "interval" : "1m"
      }
    }
  }
}
```

Notice, that "query" part is identical to data request's.
```
{
  "from" : 0,
  "size" : 150,
  "query" : {
    "bool" : {
      "must" : {
        "match_all" : { }
      },
      "filter" : {
        "bool" : {
          "must" : [ {
            "range" : {
              "timestamp" : {
                "from" : "2017-08-02 17:59:43.393",
                "to" : "2017-08-02 18:04:43.393",
                "include_lower" : true,
                "include_upper" : true
              }
            }
          }, {
            "query_string" : {
              "query" : "streams:573c6b2e1d8ed80b7a82fcb2"
            }
          } ]
        }
      }
    }
  },
  "sort" : [ {
    "timestamp" : {
      "order" : "desc"
    }
  } ]
}
```

So builder logic which builds "query" for data and hist reqeusts could be shard (imho, it must be shared).

**Comment by hc4** [comment]
also "size" option could be added to not request actual data (only hist buckets)
Got it from Grafana query builder :)
like this:
```
{
  "size" : 0,
  "query": {
    "bool": {
      "must": {
        "match_all": {}
      },
      "filter": {
        "bool": {
          "must": [{
            "range": {
              "timestamp": {
                "from": "2017-08-02 17:59:46.786",
                "to": "2017-08-02 18:04:46.786",
                "include_lower": true,
                "include_upper": true
              }
            }
          }, {
            "query_string": {
              "query": "streams:573c6b2e1d8ed80b7a82fcb2"
            }
          }]
        }
      }
    }
  },
  "aggregations": {
    "gl2_histogram": {
      "date_histogram": {
        "field": "timestamp",
        "interval": "1m"
      }
    }
  }
}
```

**Comment by kroepke** [comment]
I think #4056 should be backported to 2.3 as well, @Graylog2/engineers opinions?

**Comment by hc4** [comment]
Maybe create PR for 2.3.1 branch to not loose this fix?

**Comment by bernd** [comment]
@hc4 We decided to not back port this into 2.3 because there is a chance that this introduces some regressions. During the 2.4 release cycle we have some more time to test this.

**Comment by hc4** [comment]
Is there any ETA for 2.4?

**Comment by bernd** [comment]
@hc4 No ETA yet, but I think it will happen in Q4/2017. No promises, though. :wink:

**Comment by hc4** [comment]
Ha. Problem, with slow hist request was partially fixed in ES 5.5.2 (https://github.com/elastic/elasticsearch/pull/25872)
Now query "*" processed as match_all filter.
